autonat.go

```diff
@@ -0,0 +1,142 @@
+package autonat
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	host "github.com/libp2p/go-libp2p-host"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type NATStatus int
+
+const (
+	NATStatusUnknown NATStatus = iota
+	NATStatusPublic
+	NATStatusPrivate
+)
+
+type AutoNAT interface {
+	Status() NATStatus
+	PublicAddr() (ma.Multiaddr, error)
+}
+
+type AutoNATState struct {
+	ctx    context.Context
+	host   host.Host
+	peers  map[peer.ID]struct{}
+	status NATStatus
+	addr   ma.Multiaddr
+	mx     sync.Mutex
+}
+
+func NewAutoNAT(ctx context.Context, h host.Host) AutoNAT {
+	as := &AutoNATState{
+		ctx:    ctx,
+		host:   h,
+		peers:  make(map[peer.ID]struct{}),
+		status: NATStatusUnknown,
+	}
+
+	h.Network().Notify(as)
+	go as.background()
+
+	return as
+}
+
+func (as *AutoNATState) Status() NATStatus {
+	return as.status
+}
+
+func (as *AutoNATState) PublicAddr() (ma.Multiaddr, error) {
+	as.mx.Lock()
+	defer as.mx.Unlock()
+
+	if as.status != NATStatusPublic {
+		return nil, errors.New("NAT Status is not public")
+	}
+
+	return as.addr, nil
+}
+
+func (as *AutoNATState) background() {
+	// wait a bit for the node to come online and establish some connections
+	// before starting autodetection
+	time.Sleep(10 * time.Second)
```

Comment:
10s might be a bit short here. Automatic port forwarding has about 10s timeout.
From another hand, if it is timing out we won't get and forwarding either way.

Reply:
For testing, it might be useful for it to be configurable.

Reply:
Then again we probably do want to test the autodiscovery without waiting for this time, so maybe we can make it a constant.

We can revisit when I have the test suite ready.